python/ray/llm/_internal/batch/stages/vllm_engine_stage.py

```diff
@@ -0,0 +1,530 @@
+        self.vllm = importlib.import_module("vllm")
+
+        # Construct PoolerConfig if override_pooler_config is specified.
+        if self.task_type == "embed" and "override_pooler_config" in kwargs:
+            kwargs["override_pooler_config"] = self.vllm.config.PoolerConfig(
+                **kwargs["override_pooler_config"]
+            )
+
+        # Initialize the vLLM engine.
+        engine_args = self.vllm.AsyncEngineArgs(
+            *args,
+            **kwargs,
+            disable_log_requests=True,
+        )
+        self.engine = self.vllm.AsyncLLMEngine.from_engine_args(engine_args)
+
+        # Determine the generate function based on vLLM v0 or v1.
+        if self.vllm.envs.VLLM_USE_V1:
+            self._generate_async = self.generate_async_v1
+        else:
+            self._generate_async = self.generate_async_v0
+
+        # The performance gets really bad if there are too many requests in the pending queue.
+        # We work around it by introducing another queue that gates how many requests we are
+        # sending to the engine at once.
+        # This is not a queue of requests. Instead, this queue holds "slots". Each time
+        # we add a new request, we take one slot. When a request finishes, we add a new slot.
+        self.max_pending_requests = max_pending_requests
+        self.free_queue: asyncio.Queue[bool] = asyncio.Queue()
+        if self.max_pending_requests > 0:
+            for _ in range(self.max_pending_requests):
+                self.free_queue.put_nowait(True)
+
+    def _prepare_llm_request(self, row: Dict[str, Any]) -> LLMRequest:
+        """Prepare the inputs for LLM inference.
+
+        Args:
+            row: The row.
+
+        Returns:
+            A single LLMRequest.
+        """
+        prompt = row.pop("prompt")
+
+        if "tokenized_prompt" in row:
+            tokenized_prompt = row.pop("tokenized_prompt")
+        else:
+            tokenized_prompt = None
+
+        if "image" in row:
+            image = row.pop("image")
+        else:
+            image = []
+
+        if self.task_type == "generate":
+            params = self.vllm.SamplingParams(**row.pop("sampling_params"))
+        elif self.task_type == "embed":
+            params = self.vllm.PoolingParams()
+        else:
+            raise ValueError("Unsupported task type: %s", self.task_type)
+
+        request = self.LLMRequest(
+            request_id=self.request_id,
+            idx_in_batch=row[self.idx_in_batch_column],
+            prompt=prompt,
+            prompt_token_ids=tokenized_prompt,
+            images=image,
+            params=params,
+        )
+        self.request_id += 1
+        return request
+
+    def _parse_llm_output(self, output: Any) -> Dict[str, Any]:
+        """Parse the LLM output.
+        Args:
+            output: The LLM output.
+        Returns:
+            The parsed output.
+        """
+        # Parse the common fields.
+        output_data = {
```

Comment:
Better code organization: We should make this a dataclass / pydantic object and then create utility `from_vllm_engine_output` and then construct it. You can then call its `to_dict()` to convert it this desired dict.